bundles/org.eclipse.cdt.lsp.clangd/src/org/eclipse/cdt/lsp/internal/clangd/editor/properties/LspEditorPropertiesPage.java

```diff
@@ -9,105 +9,80 @@
  *
  * Contributors:
  * Gesa Hentschke (Bachmann electronic GmbH) - initial implementation
+ * Alexander Fedorov (ArSysOp) - rework access to preferences
  *******************************************************************************/
 
 package org.eclipse.cdt.lsp.internal.clangd.editor.properties;
 
 import java.util.Optional;
 
-import org.eclipse.cdt.lsp.internal.clangd.editor.LspEditorUiPlugin;
-import org.eclipse.cdt.lsp.internal.clangd.editor.preferences.LspEditorPreferences;
-import org.eclipse.core.resources.IProject;
+import org.eclipse.cdt.lsp.clangd.ClangdConfiguration;
+import org.eclipse.cdt.lsp.internal.clangd.ResolveProjectScope;
+import org.eclipse.cdt.lsp.internal.clangd.editor.configuration.ClangdConfigurationArea;
+import org.eclipse.core.resources.IWorkspace;
 import org.eclipse.core.resources.ProjectScope;
-import org.eclipse.core.runtime.Platform;
-import org.eclipse.core.runtime.preferences.IEclipsePreferences;
 import org.eclipse.core.runtime.preferences.IScopeContext;
-import org.eclipse.core.runtime.preferences.PreferenceMetadata;
+import org.eclipse.core.runtime.preferences.InstanceScope;
 import org.eclipse.jface.layout.GridDataFactory;
+import org.eclipse.jface.layout.GridLayoutFactory;
+import org.eclipse.jface.preference.FieldEditor;
 import org.eclipse.swt.SWT;
-import org.eclipse.swt.layout.GridData;
-import org.eclipse.swt.layout.GridLayout;
-import org.eclipse.swt.widgets.Button;
 import org.eclipse.swt.widgets.Composite;
 import org.eclipse.swt.widgets.Control;
+import org.eclipse.ui.IWorkbench;
+import org.eclipse.ui.PlatformUI;
 import org.eclipse.ui.dialogs.PropertyPage;
-import org.osgi.framework.FrameworkUtil;
-import org.osgi.service.prefs.BackingStoreException;
+import org.eclipse.ui.preferences.ScopedPreferenceStore;
 
-public class LspEditorPropertiesPage extends PropertyPage {
+public final class LspEditorPropertiesPage extends PropertyPage {
 
-	private Button preferLspEditorCheckbox;
+	private final ClangdConfiguration configuration;
+	private final IWorkspace workspace;
+	private ClangdConfigurationArea area;
+
+	public LspEditorPropertiesPage() {
+		IWorkbench workbench = PlatformUI.getWorkbench();
+		this.configuration = workbench.getService(ClangdConfiguration.class);
+		this.workspace = workbench.getService(IWorkspace.class);
+	}
 
 	@Override
 	protected Control createContents(Composite parent) {
 		Composite composite = new Composite(parent, SWT.NONE);
-		GridLayout layout = new GridLayout();
-		composite.setLayout(layout);
-		GridData data = new GridData(GridData.FILL);
-		data.grabExcessHorizontalSpace = true;
-		composite.setLayoutData(data);
-		addSettingsSection(composite);
-		load();
+		composite.setLayout(GridLayoutFactory.fillDefaults().numColumns(3).create());
+		composite.setLayoutData(GridDataFactory.fillDefaults().grab(true, false).create());
+		area = new ClangdConfigurationArea(composite, configuration.metadata());
```

Comment:
All settings on project level make only sense, when we support one LS instance per project.

Reply:
TBH, I forgot what is blocking us from having LS instance per project. Is this still a limitation of LSP4E?

Reply:
Yes. There are two issues: To enable a single LS the boolean `singleton` element in the  `org.eclipse.lsp4e.languageServer` extension point definition has to be set. To make this configurable, we could define a another language server which is a non-singleton. This solution needs an enable method for the ls definition, which is currently not implemented in the  `org.eclipse.lsp4e.languageServer` extension point, to distinguish between our singleton or non-singleton server.

Another approach would be to make the `singleton` attribute settable via the `org.eclipse.lsp4e.server.StreamConnectionProvider` or `org.eclipse.lsp4j.services.LanguageServer` interface. This is preferred by me.